File: app/src/main/java/com/vdovin/spacex/screen/detail/structure/LaunchDetailsPresenter.java
package com.vdovin.spacex.screen.detail.structure;

import com.vdovin.spacex.base.BasePresenter;
import com.vdovin.spacex.database.model.SpaceX;

import io.reactivex.disposables.Disposable;

public class LaunchDetailsPresenter extends BasePresenter<LaunchDetailsView> {

    private SpaceX spaceX;

    public boolean fullScreen;//Fix it

    public LaunchDetailsPresenter() {
    }

    public void setSpaceX(SpaceX spaceX) {
        this.spaceX = spaceX;
    }

    @Override
    protected void onLoad() {
        super.onLoad();
        getView().setupView(spaceX);
        disposables.add(back());
        disposables.add(openLink());
        disposables.add(play());
    }

    private Disposable back() {
        return getView().backButtonClicked().subscribe(o -> getView().backPressed());
    }

    private Disposable openLink() {
        return getView().linkClicked().subscribe(o -> getView().openWiki(spaceX.getWikipediaLink()));
    }

    private Disposable play() {
        return getView().launchImageClicked().flatMap(o -> getView().initializeYoutubePlayer()).subscribe(pair -> {
            fullScreen = pair.second;
            getView().playVideo(spaceX.getYoutubeVideoId(), pair.first, pair.second);
            disposables.add(detectFullScreen());
        });
    }

    private Disposable detectFullScreen() {
        return getView().detectFullscreen().subscribe(isFullscreen -> fullScreen = isFullscreen);
    }

}

File: app/src/main/java/com/vdovin/spacex/screen/details/LaunchDetailsPresenter.kt
package com.vdovin.spacex.screen.details

import android.view.View.GONE
import com.google.android.youtube.player.YouTubePlayer
import com.vdovin.spacex.base.BasePresenter
import com.vdovin.spacex.database.dao.SpaceXDao
import kotlinx.coroutines.withContext

class LaunchDetailsPresenter(private val dao: SpaceXDao) : BasePresenter<LaunchDetailsView>() {

    var flightNumber: Long? = null

    var fullscreen = false

    private var player: YouTubePlayer? = null

    override fun start() {
        flightNumber?.let {
            launch {
                val space = withContext(backgroundContext) { dao.getLaunchByFlightNumber(it) }
                space.missionName?.let { title -> view.setTitle(title) }
                space.launchDate?.let { date -> view.setDate(date) }
                space.details?.let { detail -> view.setDetails(detail) }
                space.rocketName?.let { name -> view.setRocketName(name) }
                space.payloadMass?.let { mass -> view.setPayloadMass(mass) }
                        ?: view.setPayloadMassVisibility(GONE)
                space.youtubeVideoId?.let { img ->
                    view.setImage(img)
                    view.launchImageClicked { youTubePlayer, wasRestored ->
                        if (!wasRestored) {
                            player = youTubePlayer
                            view.setLaunchImageVisibility(GONE)
                            player?.let {
                                it.loadVideo(img)
                                it.play()
                            }

                            youTubePlayer.setOnFullscreenListener {
                                fullscreen = it
                            }
                        }
                    }
                }

                view.backButtonClicked { view.backPressed() }
                space.wikipediaLink?.let { link -> view.linkClicked {  view.openWiki(link) } }
            }
        }
    }

    fun exitFullScreen(){
        player?.let {
            it.setFullscreen(false)
        }
    }
}